linear-regression/helper_functions.py

```diff
@@ -0,0 +1,367 @@
+            "mae": train_metrics[1],
+            "r2": train_metrics[2],
+        },
+        index=["train"],
+    )
+
+    test_metrics_df = pd.DataFrame(
+        {
+            "model_name": model.__class__.__name__,
+            "data": "test",
+            "rmse": np.sqrt(test_metrics[0]),
+            "mse": test_metrics[0],
+            "mae": test_metrics[1],
+            "r2": test_metrics[2],
+        },
+        index=["test"],
+    )
+
+    return pd.concat([train_metrics_df, test_metrics_df], axis=0).reset_index(drop=True)
+
+
+def backward_stepwise_train(X_train, y_train, feature_cols):
+    """
+    Perform backward stepwise selection to choose the best subset of
+    features to use for linear regression based on RMSE. This function
+    should also split the data into train and validation subsets and the RMSE should
+    be calculated based on the validation data.
+
+    Args:
+        X_train: pandas DataFrame - Input DataFrame containing all training data.
+        y_train: pandas Series - Series containing target values for training data.
+        feature_cols: list[str] - Names of feature columns to use for starting (full) model.
+
+    Returns:
+        best_idx: int - Index of the best model based on the lowest RMSE.
+        features_list: dict - Dictionary containing the list of features used for each model.
+        rmse_list: list - List of RMSE values for each model.
+    """
+
+    from sklearn.model_selection import train_test_split
+
+    # split the data into training and validation sets (xtrain, xval, ytrain, yval)
+    # use test size 0.2 and random state 1
+    xtrain, xval, ytrain, yval = train_test_split(X_train, y_train, test_size=0.2, random_state=1)
+
+    # transform the target variable using the transform_target function
+    ytrain, yval, target_transformer = transform_target(ytrain, yval)
+
+    # create an empty list to store the best models
+    best_models = []
+    
+    # create an empty list to store the RMSE values.  Also store the first set of features in it
+    remaining_features = feature_cols.copy()
+    
+    # create an empty list to store the RMSE values
+    rmse_list = []
+    
+    # create an empty dictionary to store the features used for each model
+    features_list = {}
+    
+    # create a variable k to store the number of features
+    k = len(feature_cols)
+
+    # Get full model
+    model, _, _ = linear_regression_pipeline(
+        xtrain, xval, ytrain, yval, target_transformer
+    )
+
+    # calculate the RMSE for the full model, make sure to inverse transform the target variable
+    rmse = np.sqrt(mean_squared_error(target_transformer.inverse_transform(yval.reshape(-1,1)), target_transformer.inverse_transform(model.predict(xval[feature_cols]).reshape(-1,1))))
+
+    # append the full-featured model to the best_models list
+    best_models.append(model)
+
+    # save the features in the features_list
+    features_list[0] = feature_cols
```

Comment:
feature_cols is being changed, which will change what is saved in features_list[0]